sdk/core/azure-core/inc/azure/core/operation.hpp

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include "azure/core/context.hpp"
+#include "azure/core/operation_status.hpp"
+#include "azure/core/response.hpp"
+
+#include <chrono>
+#include <string>
+#include <memory>
+
+namespace Azure { namespace Core {
+
+  //@param <T> The type of the final result of the long-running operation.
+  template <class T> class Operation {
+  private:
+    virtual std::unique_ptr<Http::RawResponse> PollInternal(Context& context) = 0;
+    virtual Response<T> PollUntilDoneInternal(std::chrono::milliseconds period, Context& context)
+        = 0;
+
+  protected:
+    OperationStatus m_status = OperationStatus::NotStarted;
+
+  public:
+    OperationStatus Status() const { return m_status; }
+
+    bool Done() const
+    {
+      return (
+          m_status == OperationStatus::Succeeded || m_status == OperationStatus::Cancelled
+          || m_status == OperationStatus::Failed);
+    }
+
+    bool HasValue() const { return (m_status == OperationStatus::Succeeded); }
+
+    virtual T Value() const = 0;
+
+    virtual std::string GetResumeToken() const = 0;
```

Comment:
Is this the equivalent of `Id` in .NET?
https://github.com/Azure/azure-sdk-for-net/blob/13aff100da135d11cd593965e3397b899ac5807d/sdk/core/Azure.Core/src/OperationOfT.cs#L25

Reply:
Yes. It's OK if we want to call it ID but token is probably a better name.

Reply:
Here's why this function is required: A machine initiates the LRO and then can get a token (or ID), pass the token(Id) to another machine. The other machine can use its XxxClient and call an overload of the StartXxx method that accepts the token/id. This overload also returns an Operation<T> and now the other machine can poll until done.

Reply:
>  A machine initiates the LRO and then can get a token (or ID), pass the token(Id) to another machine. The other machine can use its XxxClient and call an overload of the StartXxx method that accepts the token/id. This overload also returns an Operation and now the other machine can poll until done.

Can we add some unit test that verifies our `Operation<T>` behaves this way and that the `GetResumeToken` returns something meaningful, or are we planning to leave all that to the implementer of the `Operation` and the client?

Should the ID/token always be unique each time it is called (like a Guid), or be constant/stable for a specific instance of an operation?

Reply:
Makes sense. Should multiple calls to `GetResumeToken()` always return the exact same string?

Reply:
I can imagine an implementation that does always return teh same string but, in reality, they all should.